tests/containers/podmansh.pm

```diff
@@ -0,0 +1,191 @@
+
+    # verifies if running as expected user with expected groups
+    validate_script_output("$cmd id", sub { /^uid=1000\(bernhard\) gid=1000\(bernhard\) groups=1000\(bernhard\)$/ });
+
+    # verifies if correct image variant
+    validate_script_output("$cmd 'cat /etc/os-release'", sub { /^ID=\"opensuse-leap\"$/m });
+}
+
+sub prepare_user_account {
+    # make sure normal user exists
+    ensure_user_account();
+
+    # create user container
+    my $quadlet = '/usr/libexec/podman/quadlet';
+    assert_script_run("mkdir -p $systemd_user_path");
+    assert_script_run(qq(printf "$quadlet_container" >"$systemd_user_path/$unit_name"));
+    record_info('Unit', script_output("sudo -su $username $quadlet -user -v -dryrun"));
+
+    # change the default user shell to /usr/bin/podmansh
+    $initial_user_shell = script_output("getent passwd $username | cut -d':' -f7");
+    assert_script_run("usermod -s /usr/bin/podmansh $username");
+
+    # prepare ssh key for passwordless login
+    my $sshkeypath = "/root/.ssh/id_ed25519";
+    if (script_run(qq(test -e "$sshkeypath.pub")) != 0) {
+        assert_script_run(qq(ssh-keygen -t ed25519 -f "$sshkeypath" -q -N ""));
+    }
+    my $usersshpath = "/home/$username/.ssh";
+    assert_script_run(qq(mkdir -p "$usersshpath" && chown "$username:$username" "$usersshpath"));
+    assert_script_run(qq(tee -a "$usersshpath/authorized_keys" <"$sshkeypath.pub"));
+
+    # change the dafault podmansh non-root shell to match user console serial terminal regexes
+    # otherwise tty login fails
+    my $containers_conf_d = "/home/$username/.config/containers/containers.conf.d";
+    assert_script_run("mkdir -p $containers_conf_d");
+    assert_script_run("printf '[podmansh]\\nshell = \"/bin/bash\"\\n' > $containers_conf_d/podmansh.conf");
+    assert_script_run("chown -R $username:$username $containers_conf_d");
+}
+
+sub ensure_user_account {
+    # Some products don't have bernhard pre-defined (e.g. SLE Micro)
+    if (script_run("grep $username /etc/passwd") != 0) {
+        assert_script_run "useradd -m $username";
+        assert_script_run "echo '$username:$testapi::password' | chpasswd";
+        # Make sure user has access to tty group
+        my $serial_group = script_output "stat -c %G /dev/$testapi::serialdev";
+        assert_script_run "grep '^${serial_group}:.*:${username}\$' /etc/group || (chown $username /dev/$testapi::serialdev && gpasswd -a $username $serial_group)";
+    }
+}
+
+sub pre_run_hook {
+    $uid = script_output("id -u $username");
+    $systemd_user_path = "$quadlet_dir/users/$uid";
+}
+
+sub cleanup {
+    $podman->cleanup_system_host();
+
+    # disable linger and wait until user session is losed (if not already)
+    script_run("loginctl disable-linger $uid");
+    script_retry("! loginctl list-users | grep $username", retry => 5, delay => 10, timeout => 10, die => 0);
+
+    # reset user shell
+    script_run(qq(usermod -s "$initial_user_shell" "$username"));
```

Comment:
I think this is the most important cleanup/reset operation and as such I would run this first.